components/file_system/src/iosnoop/biosnoop.rs

```diff
@@ -0,0 +1,508 @@
+            let h = std::thread::spawn(move || {
+                set_io_type(IOType::Compaction);
+                let (lock, cvar) = &*pair1;
+                let mut stop = lock.lock().unwrap();
+                while !*stop {
+                    stop = cvar.wait(stop).unwrap();
+                }
+                assert_eq!(get_io_type(), IOType::Compaction);
+            });
+            handles.push(h);
+        }
+
+        // the reserved index is used, io type should be IOType::Other
+        for _ in 1..=MAX_THREAD_IDX {
+            std::thread::spawn(|| {
+                set_io_type(IOType::Compaction);
+                assert_eq!(get_io_type(), IOType::Other);
+            })
+            .join()
+            .unwrap();
+        }
+
+        {
+            let (lock, cvar) = &*pair;
+            let mut stop = lock.lock().unwrap();
+            *stop = true;
+            cvar.notify_all();
+        }
+
+        for h in handles {
+            h.join().unwrap();
+        }
+
+        // the thread indexes should be available again.
+        for _ in 1..=MAX_THREAD_IDX {
+            std::thread::spawn(|| {
+                set_io_type(IOType::Compaction);
+                assert_eq!(get_io_type(), IOType::Compaction);
+            })
+            .join()
+            .unwrap();
+        }
+    }
+
+    #[bench]
+    fn bench_write_enable_io_snoop(b: &mut Bencher) {
+        init_io_snooper().unwrap();
+        bench_write(b);
+    }
+
+    #[bench]
+    fn bench_write_disable_io_snoop(b: &mut Bencher) {
+        unsafe { BPF_TABLE = None };
+        bench_write(b);
+    }
+
+    #[bench]
+    fn bench_read_enable_io_snoop(b: &mut Bencher) {
+        init_io_snooper().unwrap();
+        bench_read(b);
+    }
+
+    #[bench]
+    fn bench_read_disable_io_snoop(b: &mut Bencher) {
+        unsafe { BPF_TABLE = None };
+        bench_read(b);
+    }
+
+    #[bench]
+    fn bench_flush_io_metrics(b: &mut Bencher) {
+        init_io_snooper().unwrap();
+        set_io_type(IOType::Write);
+
+        let tmp = TempDir::new().unwrap();
+        let file_path = tmp.path().join("bench_flush_io_metrics");
+        let mut f = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .custom_flags(O_DIRECT)
+            .open(&file_path)
+            .unwrap();
+
+        let mut w = vec![A512::default(); 1];
+        w.as_bytes_mut()[64] = 42;
+        for _ in 1..=100 {
+            f.write(w.as_bytes()).unwrap();
+        }
+        f.sync_all().unwrap();
+
+        b.iter(|| {
+            flush_io_metrics();
+        });
+    }
+
+    fn bench_write(b: &mut Bencher) {
+        let tmp = TempDir::new().unwrap();
+        let file_path = tmp.path().join("bench_write_io_snoop");
+        let mut f = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .custom_flags(O_DIRECT)
+            .open(&file_path)
+            .unwrap();
+
+        let mut w = vec![A512::default(); 1];
```

Comment:
I think we can try 1MB write without fsync, and include buffered write too.

Reply:
And btw we can start some background r/w in the meantime.

Reply:
The overhead is related to the frequency of disk io, I don't see 1MB and buffered write make any sense.

Reply:
background r/w is a little random, let's leave micro bench reproducible. And background r/w scene is covered in the real cluster bench.

Reply:
The idea is to make disk IO variance relatively small compared to total time elapsed, and also test different code paths involved in realworld IO. But I guess they aren't that import compared to cluster bench.